scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala

```diff
@@ -231,27 +225,14 @@ private[scio] final case class KryoState(kryo: Kryo,
                                          outputChunked: OutputChunked)
 
 private[scio] object KryoAtomicCoder {
-  private val logger = LoggerFactory.getLogger(this.getClass)
-
-  private[this] val defaultPoolConfig = {
-    val config = new GenericObjectPoolConfig()
-    config.setMaxTotal(Integer.MAX_VALUE)
-    config.setJmxEnabled(false)
-    config.setMaxIdle(Integer.MAX_VALUE)
-    config.setMinIdle(Integer.MAX_VALUE)
-    config
-  }
+  private val Log = LoggerFactory.getLogger(this.getClass)
+  private val Header = -1
 
-  final case class KryStatePoolFactory(f: () => KryoState)
-      extends BasePooledObjectFactory[KryoState] {
-    override def create(): KryoState = f()
+  private[this] val kryoState: ThreadLocal[KryoState] =
+    new EmptyOnDeserializationThreadLocal[KryoState]
 
-    override def wrap(obj: KryoState): PooledObject[KryoState] =
-      new DefaultPooledObject[KryoState](obj)
-  }
-
-  private[this] val kryoPool: KryoOptions => ObjectPool[KryoState] = Functions.memoize { options =>
-    val factory = KryStatePoolFactory(() => {
+  def withKryoState[R](options: KryoOptions)(f: KryoState => R): R = {
+    val ks = Option(kryoState.get()).getOrElse {
```

Comment:
Since this is a `static` var is not actually serialized, so it can be a normal `ThreadLocal`. Given this `kryoState.get()` is only null on first invocation.

Reply:
ok - that makes sense now.